mlir/include/mlir/Interfaces/TransformsInterfaces.td

```diff
@@ -0,0 +1,77 @@
+//===- TransformsInterfaces.td - Transforms interfaces -----*- tablegen -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Defines interfaces for managing transformations, including populating
+// pattern rewrites.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef MLIR_INTERFACES_TRANSFORMSINTERFACES_TD
+#define MLIR_INTERFACES_TRANSFORMSINTERFACES_TD
+
+include "mlir/IR/OpBase.td"
+
+//===----------------------------------------------------------------------===//
+// Conversion patterns attribute interface
+//===----------------------------------------------------------------------===//
+
+def ConversionPatternsAttrInterface :
+    AttrInterface<"ConversionPatternsAttrInterface"> {
+  let description = [{
+    This interfaces allows using attributes to configure the dialect conversion
+    infrastructure, this includes:
+     - The conversion target.
+     - The type converter.
+     - The pattern set.
+    
+    The conversion target and type converter are passed through the
+    `ConversionPatternAttrOptions` class. Passing them through this class
+    and by reference allows sub-classing the base option class, allowing
+    specializations like `LLVMConversionPatternAttrOptions` for converting to
+    LLVM.
+  }];
+  let cppNamespace = "::mlir";
+  let methods = [
+    InterfaceMethod<
+      /*desc=*/[{
+        Populate the dialect conversion target, type converter and pattern set.
+      }],
+      /*retTy=*/"void",
+      /*methodName=*/"populateConversionPatterns",
+      /*args=*/(ins "::mlir::ConversionPatternAttrOptions&":$options,
+                    "::mlir::RewritePatternSet&":$patternSet)>
+  ];
+}
```

Comment:
I don't understand what this interface provides? 
"conversion" is very context dependent: for example a math operation can have different conversion for going to LLVM, to SPIRV, or to TensorFlow for example.

Reply:
It provides the ability to configure the dialect conversion data structures, regardless if the target is LLVM, SPIR-V, or things like `scf` to `cf`.

I debated with myself whether scope it to a `ConvertToLLVM` interface, or let it have a more generic purpose if someone else found it useful, and avoid recreating this interface for every possible dialect conversion.

I choose the generic version. However, I don't have strong opinions for it, I can switch it to a `ConvertToLLVM` specific interface.

Reply:
> regardless if the target is LLVM, SPIR-V, or things like scf to cf. 

Right: I don't quite understand this. How can it be useful to have this interface if you don't know the target? The patterns to populate for an operation are entirely dependent of the conversion you're doing aren't they?

Reply:
Yes.
That's precisely why I pass `ConversionPatternAttrOptions` by reference, to sub-class it and allow interfaces to discern which conversion to populate. But also at the same time, not forcing every conversion to add an interface for it. See for example the `NVVMTargetPatternsAttrInterface::populateConversionPatterns`:
```c++
void NVVMTargetPatternsAttrInterface::populateConversionPatterns(
    Attribute attr, ConversionPatternAttrOptions &options,
    RewritePatternSet &patterns) const {
  auto *llvmOptions = dyn_cast<LLVMConversionPatternAttrOptions>(&options);
  // Bail if the options are invalid.
  if (!llvmOptions)
    return;
  configureGpuToNVVMConversionLegality(options.getConversionTarget());
  configureGpuToNVVMTypeConverter(llvmOptions->getLLVMTypeConverter());
  populateGpuToNVVMConversionPatterns(llvmOptions->getLLVMTypeConverter(),
                                      patterns);
}
```
However, I see 100% the point of having interfaces for each conversion. I can change it to an LLVM specific one.